File: search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>&A, int key) {
        int low=0,high=A.size()-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(A[mid]==key) return 1;
            if(A[low]==A[mid] && A[mid]==A[high]) {
                low++; high--; continue;
            }
            if(A[low]<=A[mid]){
                if(key>=A[low] && key<A[mid]) high=mid-1;
                else low=mid+1;
            }
            else{
                if(key>A[mid] && key<=A[high]) low=mid+1;
                else high=mid-1;
            }
        }
        return 0;
    }
};

File: search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        n=len(nums)
        low=0
        high=n-1
        while(low<=high):
            mid = low+(high-low)//2
            if(nums[mid]==target): return True
            if(nums[low]==nums[mid] and nums[mid]==nums[high]):
                low+=1
                high-=1
                continue
            if(nums[low]<=nums[mid]):
                if(nums[low]<=target and target<=nums[mid]): high=mid
                else: low=mid+1
            else:
                if(nums[mid]<=target and target<=nums[high]): low=mid
                else: high=mid-1
        return False